library/core/resource_loader/src/main/java/org/quiltmc/qsl/resource/loader/api/InMemoryResourcePack.java

```diff
@@ -0,0 +1,297 @@
+import org.quiltmc.qsl.base.api.util.TriState;
+
+/**
+ * Represents an in-memory resource pack.
+ * <p>
+ * The resources of this pack are stored in memory instead of it being on-disk.
+ */
+public abstract class InMemoryResourcePack implements MutableResourcePack {
+	private static final Logger LOGGER = LogUtils.getLogger();
+	private static final ExecutorService EXECUTOR_SERVICE;
+	private static final boolean DUMP = TriState.fromProperty("quilt.resource_loader.debug.pack.dump_from_in_memory")
+			.toBooleanOrElse(QuiltLoader.isDevelopmentEnvironment());
+	private final Map<Identifier, Supplier<byte[]>> assets = new ConcurrentHashMap<>();
+	private final Map<Identifier, Supplier<byte[]>> data = new ConcurrentHashMap<>();
+	private final Map<String, Supplier<byte[]>> root = new ConcurrentHashMap<>();
+
+	@Override
+	public @Nullable InputStream openRoot(String fileName) {
+		if (!fileName.contains("/") && !fileName.contains("\\")) {
+			return this.openResource(this.root, fileName);
+		} else {
+			throw new IllegalArgumentException("Root resources can only be filenames, not paths (no / allowed!)");
+		}
+	}
+
+	@Override
+	public InputStream open(ResourceType type, Identifier id) throws IOException {
+		var stream = this.openResource(this.getResourceMap(type), id);
+
+		if (stream == null) {
+			throw new FileNotFoundException("Could not find resource \"" + id + "\" (" + type.getDirectory() + ") in pack " + this.getName() + ".");
+		}
+
+		return stream;
+	}
+
+	protected @Nullable <T> InputStream openResource(Map<T, Supplier<byte[]>> map, @NotNull T key) {
+		var supplier = map.get(key);
+
+		if (supplier == null) {
+			return null;
+		}
+
+		byte[] bytes = supplier.get();
+
+		if (bytes == null) {
+			return null;
+		}
+
+		return new ByteArrayInputStream(bytes);
+	}
+
+	@Override
+	public Collection<Identifier> findResources(ResourceType type, String namespace, String startingPath, Predicate<Identifier> pathFilter) {
+		return this.getResourceMap(type).keySet().stream()
+				.filter(id -> id.getNamespace().equals(namespace) && id.getPath().startsWith(startingPath))
+				.filter(pathFilter)
+				.collect(Collectors.toList());
+	}
+
+	@Override
+	public boolean contains(ResourceType type, Identifier id) {
+		return this.getResourceMap(type).containsKey(id);
+	}
+
+	@Override
+	public @Unmodifiable Set<String> getNamespaces(ResourceType type) {
+		return this.getResourceMap(type).keySet().stream()
+				.map(Identifier::getNamespace)
+				.collect(Collectors.toUnmodifiableSet());
+	}
+
+	@Override
+	public <T> @Nullable T parseMetadata(ResourceMetadataReader<T> metaReader) throws IOException {
+		if (!this.root.containsKey(ResourcePack.PACK_METADATA_NAME)) {
+			var json = new JsonObject();
+			var packJson = new JsonObject();
+			packJson.addProperty("description", "A virtual resource pack.");
+			packJson.addProperty("pack_format", 5); // This is like, not read by any significant system when invisible to users.
```

Comment:
i know that this isn't being read by anyone here, but
```suggestion
			packJson.addProperty("pack_format", 10); // This is like, not read by any significant system when invisible to users.
```
pretty please? :P

Reply:
I chose `5` because it's the smallest number that doesn't have special treatment by MC.
This feels very arbitrary.

Reply:
oh, I see; I mentioned 10 'cause I imagined that the current number was going to be the one not subject to things such as DFU

Reply:
Isn't there a constant for the latest resource pack version? I'd use that to always match the latest MC.

Reply:
The issue is the pack version is side dependent, data packs and resource packs don't have the same versions, making it impossible to pair them up as one value.